vllm-model-provider/server/server.go

```diff
@@ -0,0 +1,124 @@
+)
+
+func Run(apiKey, port string) error {
+    endpoint := os.Getenv("OBOT_VLLM_MODEL_PROVIDER_ENDPOINT")
+    if endpoint == "" {
+        return fmt.Errorf("OBOT_VLLM_MODEL_PROVIDER_ENDPOINT environment variable not set")
+    }
+
+    // Remove https:// or http:// if present
+    endpoint = strings.TrimPrefix(endpoint, "https://")
+    endpoint = strings.TrimPrefix(endpoint, "http://")
+
+    mux := http.NewServeMux()
+
+    s := &server{
+        apiKey:   apiKey,
+        port:     port,
+        endpoint: endpoint,
+    }
+
+    mux.HandleFunc("/{$}", s.healthz)
+    mux.Handle("GET /v1/models", &httputil.ReverseProxy{
+        Director:       s.proxy,
+        ModifyResponse: s.rewriteModelsResponse,
+        Transport: &http.Transport{
+            TLSClientConfig: &tls.Config{
+                InsecureSkipVerify: true,
+            },
+        },
+    })
+    mux.Handle("/{path...}", &httputil.ReverseProxy{
+        Director: s.proxy,
+        Transport: &http.Transport{
+            TLSClientConfig: &tls.Config{
+                InsecureSkipVerify: true,
+            },
+        },
+    })
+
+    httpServer := &http.Server{
+        Addr:    "127.0.0.1:" + port,
+        Handler: mux,
+    }
+
+    if err := httpServer.ListenAndServe(); !errors.Is(err, http.ErrServerClosed) {
+        return err
+    }
+
+    return nil
+}
+
+type server struct {
+    apiKey, port, endpoint string
+}
+
+func (s *server) healthz(w http.ResponseWriter, _ *http.Request) {
+    _, _ = w.Write([]byte("http://127.0.0.1:" + s.port))
+}
+
+func (s *server) rewriteModelsResponse(resp *http.Response) error {
+    if resp.StatusCode != http.StatusOK {
+        return nil
+    }
+
+    originalBody := resp.Body
+    defer originalBody.Close()
+
+    if resp.Header.Get("Content-Encoding") == "gzip" {
+        var err error
+        originalBody, err = gzip.NewReader(originalBody)
+        if err != nil {
+            return fmt.Errorf("failed to create gzip reader: %w", err)
+        }
+        defer originalBody.Close()
+        resp.Header.Del("Content-Encoding")
+    }
+
+    var models openai.ModelsList
+    if err := json.NewDecoder(originalBody).Decode(&models); err != nil {
+        return fmt.Errorf("failed to decode models response: %w, %d, %v", err, resp.StatusCode, resp.Header)
+    }
+
+    // Set all models as LLM
+    for i, model := range models.Models {
+        if model.Metadata == nil {
+            model.Metadata = make(map[string]string)
+        }
+        model.Metadata["usage"] = "llm"
+        models.Models[i] = model
+    }
+
+    b, err := json.Marshal(models)
+    if err != nil {
+        return fmt.Errorf("failed to marshal models response: %w", err)
+    }
+
+    resp.Body = io.NopCloser(bytes.NewReader(b))
+    resp.Header.Set("Content-Length", fmt.Sprintf("%d", len(b)))
+    return nil
+}
+
+func (s *server) proxy(req *http.Request) {
+    req.URL.Host = s.endpoint
+    req.URL.Scheme = "https"
```

Comment:
I noticed that we are forcing this to be HTTPS here (and that InsecureSkipVerify gets set on the TLS config for the reverse proxy). I'm curious to know why this is. Since vLLM is something you run yourself, I could see people running it on a local network over just HTTP, or running it on a larger scale and using HTTPS (and wanting to keep verification on).

Reply:
fair point - ill default to HTTP for localhost/127.0.0.1 and HTTPS for all other hosts
fwiw i dont think users will run obot on the same machine with their GPUs